src/Vulkan/RHI/VkPipelineState.cpp

```diff
@@ -22,12 +81,174 @@ void VkGraphicsPipelineState::Compile(const Shader& vertexShader,
+        .back = GraphicsPipeline_Internal::StencilOpStateToVkStencilOpState(key.depthStencilState.back),
+        .minDepthBounds = key.depthStencilState.minDepthBounds,
+        .maxDepthBounds = key.depthStencilState.maxDepthBounds,
+    };
+
+    std::vector<::vk::PipelineColorBlendAttachmentState> colorAttachments(key.colorBlendState.attachments.size());
+    std::ranges::transform(
+        key.colorBlendState.attachments,
+        colorAttachments.begin(),
+        [](const ColorBlendState::ColorBlendAttachment& attachment)
+        {
+            return ::vk::PipelineColorBlendAttachmentState{
+                .blendEnable = attachment.blendEnabled,
+                .srcColorBlendFactor = static_cast<::vk::BlendFactor>(attachment.srcColorBlendFactor),
+                .dstColorBlendFactor = static_cast<::vk::BlendFactor>(attachment.dstColorBlendFactor),
+                .colorBlendOp = static_cast<::vk::BlendOp>(attachment.colorBlendOp),
+                .srcAlphaBlendFactor = static_cast<::vk::BlendFactor>(attachment.srcAlphaBlendFactor),
+                .dstAlphaBlendFactor = static_cast<::vk::BlendFactor>(attachment.dstAlphaBlendFactor),
+                .alphaBlendOp = static_cast<::vk::BlendOp>(attachment.alphaBlendOp),
+                .colorWriteMask = static_cast<::vk::ColorComponentFlags>(attachment.colorWriteMask)
+            };
+        });
+
+    ::vk::PipelineColorBlendStateCreateInfo blendStateCI{
+        .logicOpEnable = key.colorBlendState.logicOpEnabled,
+        .logicOp = static_cast<::vk::LogicOp>(key.colorBlendState.logicOp),
+        .attachmentCount = static_cast<u32>(colorAttachments.size()),
+        .pAttachments = colorAttachments.data(),
+        .blendConstants = key.colorBlendState.blendConstants,
+    };
+
+    std::array dynamicStates = { ::vk::DynamicState::eViewportWithCount,
+                                 ::vk::DynamicState::eScissorWithCount,
+                                 ::vk::DynamicState::ePrimitiveTopology,
+                                 ::vk::DynamicState::ePrimitiveRestartEnable };
+
+    ::vk::PipelineDynamicStateCreateInfo dynamicStateInfo{
+        .dynamicStateCount = static_cast<uint32_t>(dynamicStates.size()),
+        .pDynamicStates = dynamicStates.data(),
+    };
+
+    std::vector<::vk::Format> attachmentFormats(key.renderTargetState.colorFormats.size());
+    std::ranges::transform(key.renderTargetState.colorFormats, attachmentFormats.begin(), TextureFormatToVulkan);
+
+    const ::vk::PipelineRenderingCreateInfoKHR pipelineRenderingCI{
+        .colorAttachmentCount = static_cast<u32>(attachmentFormats.size()),
+        .pColorAttachmentFormats = attachmentFormats.data(),
+    };
+
+    ::vk::PipelineViewportStateCreateInfo viewportStateCI{
+        .viewportCount = 0,
+        .pViewports = nullptr,
+        .scissorCount = 0,
+        .pScissors = nullptr,
+    };
+
+    ::vk::GraphicsPipelineCreateInfo graphicsPipelineCI{ .pNext = &pipelineRenderingCI,
```

Comment:
I didn't check in detail this section since its a LOT of boilerplate conversion/setup code.
The main things to know are that invalid states should be checked at some point on our side (I believe the DX12-side has a validation function inside CreateGraphicsPipelineState in its RHI, something similar for the most egregious errors could be nice.
Also MSAA is unsupported (so any attempt to use it should error out).